eth-providers/src/base-provider.ts

```diff
@@ -1459,39 +1459,41 @@ export abstract class BaseProvider extends AbstractProvider {
             return logger.throwError('header not found', PROVIDER_ERRORS.HEADER_NOT_FOUND, { blockNumber });
           }
 
-          blockHash = _blockHash.toHex();
+          const blockHash = _blockHash.toHex();
 
           if (isFinalized) {
             this.storageCache.set(cacheKey, _blockHash.toU8a());
           }
-        }
 
-        if (!blockHash) {
-          return logger.throwArgumentError('blocktag should be a hex string or number', 'blockTag', blockTag);
+          return blockHash;
         }
 
-        return blockHash;
+        return logger.throwArgumentError(
+          'blocktag should be number | hex string | \'latest\' | \'earliest\' | \'finalized\' | \'safe\'',
+          'blockTag',
+          blockTag
+        );
       }
     }
   };
 
+  _isBlockCanonical = async (blockHash: string, _blockNumber?: number): Promise<boolean> => {
```

Comment:
Why is `_blockNumber?` here?

Reply:
```
const blockNumber = _blockNumber ?? await this._getBlockNumber(blockHash);
```
if we already know the block number (which is the case for receipt), we can save an async call